editoast/src/views/timetable/train_schedule.rs

```diff
@@ -564,6 +574,147 @@ async fn occupancy_blocks(
     Ok(Json(occupancy_blocks_result))
 }
 
+#[derive(Debug, Default, Clone, Serialize, Deserialize, ToSchema)]
+struct TrackOccupancyForm {
+    train_schedule_ids: Vec<i64>,
+    op_id: String,
+    infra_id: i64,
+    electrical_profile_set_id: Option<i64>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
+struct TrackOccupancy {
+    time_begin: DateTime<Utc>,
+    #[schema(value_type = chrono::Duration, example = "PT5M")]
+    duration: PositiveDuration,
+    train_schedule_id: i64,
+}
+
+#[utoipa::path(
+    post, path = "",
+    tag = "train_schedule",
+    request_body = inline(TrackOccupancyForm),
+    responses(
+        (status = 200, description = "Track section occupancy periods for a set of train schedules", body = inline(HashMap<String, Vec<TrackOccupancy>>)),
+    ),
+)]
+async fn track_occupancy(
+    State(AppState {
+        db_pool,
+        valkey,
+        core_client,
+        ..
+    }): State<AppState>,
+    Extension(_auth): AuthenticationExt,
+    Json(TrackOccupancyForm {
+        train_schedule_ids,
+        op_id,
+        infra_id,
+        electrical_profile_set_id,
+    }): Json<TrackOccupancyForm>,
+) -> Result<Json<HashMap<String, Vec<TrackOccupancy>>>> {
+    let mut result: HashMap<String, Vec<TrackOccupancy>> = HashMap::new();
+    // Retrieve infra / train_schedules / operational points
+    let infra = Infra::retrieve_real_or_fail(db_pool.get().await?, infra_id, || {
+        TrainScheduleError::InfraNotFound { infra_id }
+    })
+    .await?;
+
+    let train_schedules: Vec<models::TrainSchedule> =
+        models::TrainSchedule::retrieve_batch_or_fail(
+            &mut db_pool.get().await?,
+            train_schedule_ids,
+            |missing| TrainScheduleError::BatchTrainScheduleNotFound {
+                number: missing.len(),
+            },
+        )
+        .await?;
+
+    let operational_point = OperationalPointModel::retrieve_real_or_fail(
+        db_pool.get().await?,
+        (infra_id, op_id.clone()),
+        || TrainScheduleError::OperationalPointNotFound { op_id },
+    )
+    .await?
+    .schema;
+
+    // Retrieve simulations
+    let simulations_result = train_simulation_batch(
+        &mut db_pool.get().await?,
+        valkey,
+        core_client,
+        &train_schedules,
+        &infra,
+        electrical_profile_set_id,
+    )
+    .await?;
+    let operational_point_track_offsets = operational_point.track_offset();
+
+    for ((simulation, pathfinding), train_schedule) in
+        simulations_result.iter().zip(train_schedules)
+    {
+        let track_ranges = match pathfinding {
+            PathfindingResult::Success(pathfinding_result_success) => {
+                &pathfinding_result_success.track_section_ranges
+            }
+            PathfindingResult::Failure(_) => continue,
+        };
+        let mut abs_offset = 0;
+        for track_range in track_ranges {
```

Comment:
This loop contains some algorithm. It might be nice to extract it in a function, and add unit tests for it. Note that in case of not obvious algorithm, multiple unit tests with edge cases helps documenting it (it shows what we expect from the algorithm, some sort of specification), so it's not only about ensuring that it works correctly.